Implement the `require-trusted-types-for` pre-navigation check

lukewarlow opened the pull request:
#### c28bce0ecd22a5a41b3b0f9c8958be13880e5f64
<pre>
Implement the `require-trusted-types-for` pre-navigation check
<a href="https://bugs.webkit.org/show_bug.cgi?id=267695">https://bugs.webkit.org/show_bug.cgi?id=267695</a>

Reviewed by Youenn Fablet.

When the &apos;require-trusted-types-for&apos; CSP directive is present javascript URLs are now passed through a default policy
or rejected if one doesn&apos;t exist.

Spec: <a href="https://w3c.github.io/trusted-types/dist/spec/#require-trusted-types-for-pre-navigation-check">https://w3c.github.io/trusted-types/dist/spec/#require-trusted-types-for-pre-navigation-check</a>

* LayoutTests/imported/w3c/web-platform-tests/trusted-types/support/navigation-report-only-support.html:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/support/navigation-support.html:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-navigation-expected.txt:
* Source/WebCore/bindings/js/ScriptController.cpp:
(WebCore::ScriptController::executeJavaScriptURL):
* Source/WebCore/dom/TrustedType.cpp:
(WebCore::requireTrustedTypesForPreNavigationCheckPasses):
* Source/WebCore/dom/TrustedType.h:

Canonical link: <a href="https://commits.webkit.org/275687@main">https://commits.webkit.org/275687@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f19041a00b25339d59607d601d13d4d47a5faf0d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/42475 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/21493 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44869 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/45076 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/38592 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/44782 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/24723 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/18844 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/35174 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43049 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/18407 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/36562 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16114 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/16055 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/37607 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/46549 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/38645 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37931 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/41847 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/17291 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/14226 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/40476 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/18910 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9509 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/18972 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/18555 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

lukewarlow commented on LayoutTests/imported/w3c/web-platform-tests/trusted-types/support/navigation-support.html:
```diff
@@ -32,7 +32,7 @@
   // we'll only click if we don't find fragment in the current URL.
   const target_script = `location.href='${location.href}&continue';`;
-  const target = `javascript:"<script>${target_script}</scri${""}pt>"`;
+  const target = `javascript:${target_script}`;
```
https://github.com/web-platform-tests/wpt/pull/44583

WebKit doesn't support JavaScript URLs containing `<script>` tags so this test fails within WebKit regardless of what it should actually be testing. I'm unaware why the test uses this odd construction in the first place. But this removes it.

youennf replied:
Given https://github.com/web-platform-tests/wpt/pull/44583 has been reviewed, you can merge it there already. Let me know if you need help there.

youennf commented on Source/WebCore/bindings/js/ScriptController.cpp:
```diff
@@ -824,16 +825,24 @@ void ScriptController::executeJavaScriptURL(const URL& url, RefPtr<SecurityOrigi
-
     JSDOMGlobalObject* globalObject = jsWindowProxy(mainThreadNormalWorld()).window();
+
+    String urlString(url.string());
```
Why not inlining at call site, that might reduce count churn.

lukewarlow replied:
It needs to be a local variable because it can be mutated by the preNavigationCheck and then is passed to allowJavaScriptURLs.

Based on https://w3c.github.io/trusted-types/dist/spec/#require-trusted-types-for-pre-navigation-check the URL needs to be changed for this later check.

youennf commented on Source/WebCore/bindings/js/ScriptController.cpp:
```diff
@@ -824,16 +825,24 @@ void ScriptController::executeJavaScriptURL(const URL& url, RefPtr<SecurityOrigi
     JSDOMGlobalObject* globalObject = jsWindowProxy(mainThreadNormalWorld()).window();
+
+    String urlString(url.string());
+    auto preNavigationCheckPasses = requireTrustedTypesForPreNavigationCheckPasses(*globalObject->scriptExecutionContext(), urlString);
```
Are we sure scriptExecutionContext() is not null?

lukewarlow replied:
It should be but I'll add a check just to cover that scenario.

youennf commented on Source/WebCore/bindings/js/ScriptController.cpp:
```diff
@@ -824,16 +825,24 @@ void ScriptController::executeJavaScriptURL(const URL& url, RefPtr<SecurityOrigi
+    if (preNavigationCheckPasses.hasException() || !preNavigationCheckPasses.releaseReturnValue())
+        return;
+
+    if (!ownerDocument->checkedContentSecurityPolicy()->allowJavaScriptURLs(ownerDocument->url().string(), eventHandlerPosition().m_line, urlString, nullptr))
```
What is guaranting ownerDocument to be non null?

lukewarlow replied:
It would have crashed earlier in this function if it weren't? This specific code is pre-existing just moved.

youennf commented on Source/WebCore/dom/TrustedType.cpp:
```diff
@@ -168,4 +168,36 @@ ExceptionOr<String> trustedTypeCompliantString(TrustedType expectedType, ScriptE
 }
 
+// https://w3c.github.io/trusted-types/dist/spec/#require-trusted-types-for-pre-navigation-check
+ExceptionOr<bool> requireTrustedTypesForPreNavigationCheckPasses(ScriptExecutionContext& scriptExecutionContext, String& urlString)
```
Can urlString be const&, this would make this function easier to read.
I would tend to return an URL here, and call site would check URL.isValid() as Blocked/Allowed

youennf reviewed (commented):
LGTM overall, a few comments below.

youennf commented on Source/WebCore/bindings/js/ScriptController.cpp:
```diff
@@ -824,16 +825,30 @@ void ScriptController::executeJavaScriptURL(const URL& url, RefPtr<SecurityOrigi
-
     JSDOMGlobalObject* globalObject = jsWindowProxy(mainThreadNormalWorld()).window();
+
+    if (!globalObject->scriptExecutionContext())
```
Let's take a RefPtr.

youennf commented on Source/WebCore/dom/TrustedType.cpp:
```diff
@@ -168,4 +168,40 @@ ExceptionOr<String> trustedTypeCompliantString(TrustedType expectedType, ScriptE
+        if (allowMissingTrustedTypes)
+            return URL(urlString);
+
+        return Exception { ExceptionCode::TypeError, makeString("This assignment requires a ", trustedTypeToString(expectedType)) };
```
I would tend to do `if (!allowMissingTrustedTypes) return Exception...`

lukewarlow replied:
I was between the two so will swap it

lukewarlow committed:
Implement the `require-trusted-types-for` pre-navigation check
https://bugs.webkit.org/show_bug.cgi?id=267695

Reviewed by Youenn Fablet.

When the 'require-trusted-types-for' CSP directive is present javascript URLs are now passed through a default policy
or rejected if one doesn't exist.

Spec: https://w3c.github.io/trusted-types/dist/spec/#require-trusted-types-for-pre-navigation-check

* LayoutTests/imported/w3c/web-platform-tests/trusted-types/support/navigation-report-only-support.html:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/support/navigation-support.html:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-navigation-expected.txt:
* Source/WebCore/bindings/js/ScriptController.cpp:
(WebCore::ScriptController::executeJavaScriptURL):
* Source/WebCore/dom/TrustedType.cpp:
(WebCore::requireTrustedTypesForPreNavigationCheckPasses):
* Source/WebCore/dom/TrustedType.h:

Canonical link: https://commits.webkit.org/275687@main
